File: frontend/src/lib/types/call-state.ts
export interface CallState {
  currentCall: {
    callAttemptId: string;
    sourceId: string;
    state: 'incoming' | 'connected' | 'ended';
    startTime: number;
    duration: number;
    callType: 'voice' | 'video';
  } | null;
  webrtc: {
    peerConnection: RTCPeerConnection | null;
    localStream: MediaStream | null;
    remoteStream: MediaStream | null;
    connectionState: RTCPeerConnectionState;
  };
  ui: {
    status: 'available' | 'unavailable' | 'busy' | 'ringing';
    showIncomingCallModal: boolean;
    showCallControls: boolean;
    isMuted: boolean;
  };
  callHistory: CallRecord[];
}

export interface CustomerCallState {
  callAttemptId: string | null;
  state: 'idle' | 'connecting' | 'ringing' | 'connected' | 'ended' | 'failed';
  callType: 'voice' | 'video';
  webrtc: {
    peerConnection: RTCPeerConnection | null;
    localStream: MediaStream | null;
    remoteStream: MediaStream | null;
  };
  ui: {
    showCallButton: boolean;
    showCallControls: boolean;
    statusMessage: string;
    isMuted: boolean;
    isVideoEnabled: boolean;
  };
  handle: string;
  sourceId: string;
}

export interface CallRecord {
  callAttemptId: string;
  sourceId: string;
  startTime: number;
  endTime: number;
  duration: number;
  device: 'web' | 'android';
  status: 'completed' | 'missed' | 'failed' | 'timeout' | 'cancelled';
}

File: flutter/lib/src/call/call_state.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../protocol/protocol.dart';

part 'call_state.freezed.dart';

/// Call session data
@freezed
class CallSession with _$CallSession {
  const factory CallSession({
    required String callAttemptId,
    required String handle,
    required CallType callType,
    required CallState state,
    String? sourceId,
    MediaCapabilities? localCapabilities,
    MediaCapabilities? remoteCapabilities,
    DateTime? startTime,
    DateTime? connectTime,
    DateTime? endTime,
    CallEndReason? endReason,
    RTCSessionDescriptionInit? localOffer,
    RTCSessionDescriptionInit? remoteOffer,
    RTCSessionDescriptionInit? localAnswer,
    RTCSessionDescriptionInit? remoteAnswer,
    @Default([]) List<RTCIceCandidate> localCandidates,
    @Default([]) List<RTCIceCandidate> remoteCandidates,
    @Default(true) bool isVideoEnabled,
  }) = _CallSession;

  const CallSession._();

  /// Duration of call
  Duration? get duration {
    if (startTime == null) return null;
    final end = endTime ?? DateTime.now();
    return end.difference(startTime!);
  }

  /// Is call active (connected or connecting)
  bool get isActive {
    return state == CallState.connecting ||
        state == CallState.connected ||
        state == CallState.ringing;
  }

  /// Is call ended
  bool get isEnded {
    return state == CallState.ended ||
        state == CallState.failed ||
        state == CallState.cancelled ||
        state == CallState.busy ||
        state == CallState.unavailable ||
        state == CallState.timeout;
  }
}

/// Call manager state
@freezed
class CallManagerState with _$CallManagerState {
  const factory CallManagerState({
    CallSession? currentCall,
    @Default([]) List<CallSession> callHistory,
    @Default(false) bool isAvailable,
    String? deviceId,
    DeviceType? deviceType,
  }) = _CallManagerState;

  const CallManagerState._();

  /// Has active call
  bool get hasActiveCall => currentCall?.isActive ?? false;

  /// Can initiate new call
  bool get canInitiateCall => !hasActiveCall && isAvailable;

  /// Can accept incoming call
  bool get canAcceptCall =>
      currentCall?.state == CallState.ringing && isAvailable;
}
